python/lsst/drp/tasks/wcsfit.py

```diff
@@ -0,0 +1,1019 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <https://www.lsstcorp.org/LegalNotices/>.
+#
+import numpy as np
+import astropy.time
+import astropy.units as u
+import astropy.coordinates
+import yaml
+import wcsfit
+import pandas as pd
+import astshim
+
+import lsst.geom
+import lsst.pex.config as pexConfig
+import lsst.pipe.base as pipeBase
+import lsst.sphgeom
+import lsst.afw.table
+import lsst.afw.geom as afwgeom
+from lsst.meas.algorithms import ReferenceObjectLoader, ReferenceSourceSelectorTask
+from lsst.meas.algorithms.sourceSelector import sourceSelectorRegistry
+
+__all__ = ["WCSFitConnections", "WCSFitConfig", "WCSFitTask"]
+
+
+def lookupVisitRefCats(datasetType, registry, quantumDataId, collections):
+    """Lookup function that finds all refcats for all visits that overlap a
+    tract, rather than just the refcats that directly overlap the tract.
+    Borrowed from jointcal.
+    Parameters
+    ----------
+    datasetType : `lsst.daf.butler.DatasetType`
+        Type of dataset being searched for.
+    registry : `lsst.daf.butler.Registry`
+        Data repository registry to search.
+    quantumDataId : `lsst.daf.butler.DataCoordinate`
+        Data ID of the quantum; expected to be something we can use as a
+        constraint to query for overlapping visits.
+    collections : `Iterable` [ `str` ]
+        Collections to search.
+    Returns
+    -------
+    refs : `Iterator` [ `lsst.daf.butler.DatasetRef` ]
+        Iterator over refcat references.
+    """
+    refs = set()
+    # Use .expanded() on the query methods below because we need data IDs with
+    # regions, both in the outer loop over visits (queryDatasets will expand
+    # any data ID we give it, but doing it up-front in bulk is much more
+    # efficient) and in the data IDs of the DatasetRefs this function yields
+    # (because the RefCatLoader relies on them to do some of its own
+    # filtering).
+    for visit_data_id in set(registry.queryDataIds("visit", dataId=quantumDataId).expanded()):
+        refs.update(
+            registry.queryDatasets(
+                datasetType,
+                collections=collections,
+                dataId=visit_data_id,
+                findFirst=True,
+            ).expanded()
+        )
+    yield from refs
+
+
+def makeRefCovarianceMatrix(refCat, inputUnit=u.radian, outputCoordUnit=u.degree, outputPMUnit=u.marcsec):
+    """Make a covariance matrix for the reference catalog including proper
+    motion and parallax.
```

Comment:
I think this docstring needs some more explanation, because it looks like you're making a very specific kind of flattened covariance matrix. I assume this is specifically formatted for gbdes?

You also don't need `inputUnit`, since the refcat columns already have units attached.

Reply:
The covariance matrix is just a np.ravel(order='C')-style flattening. I added to the documentation.

I've had some trouble before pulling the units out of the `simpleCatalog`, I finally figured out I could do
```
unit = refCat.schema[<key>].asField().getUnits()
```
However, this only works for the keys where the units are listed when you print the schema. For 'coord_ra', for example, when you look at the table you see 'rad' under 'coord_ra' in the column name, so obviously the information is there, but I can't figure out how to access this.

Reply:
You just have to "know" that the `coord_ra` and `coord_dec` are in radians if you are using the array access rather than the unbelievably slow row-by-row access.  But note that that statement only applies to afw tables (like the refcats); for parquet tables sometimes they're in degrees.